caffeine/src/main/java/com/github/benmanes/caffeine/cache/LocalCacheFactory.java

```diff
@@ -80,10 +83,21 @@ static String getClassName(Caffeine<?, ?> builder) {
 
   static <K, V> BoundedLocalCache<K, V> loadFactory(Caffeine<K, V> builder,
       @Nullable AsyncCacheLoader<? super K, V> cacheLoader, boolean async, String className) {
+    var constructor = CONSTRUCTORS.get(className);
+    if (constructor == null) {
+      constructor = CONSTRUCTORS.computeIfAbsent(className, LocalCacheFactory::newConstructor);
+    }
+    try {
+      return (BoundedLocalCache<K, V>) constructor.invoke(builder, cacheLoader, async);
+    } catch (Throwable t) {
+      throw new IllegalStateException(className, t);
+    }
+  }
+
+  static MethodHandle newConstructor(String className) {
     try {
-      Class<?> clazz = Class.forName(className);
-      MethodHandle handle = LOOKUP.findConstructor(clazz, FACTORY);
-      return (BoundedLocalCache<K, V>) handle.invoke(builder, cacheLoader, async);
+      Class<?> clazz = Class.forName(LocalCacheFactory.class.getPackageName() + "." + className);
+      return LOOKUP.findConstructor(clazz, FACTORY);
```

Comment:
Change to: `return LOOKUP.findConstructor(clazz, FACTORY).changeReturnType(BoundedLocalCache.class)`

This will adapt the return type, so the above call will have correct typoeninformation and invokeExact works.

Reply:
One additional thing: Since Java 11 I'd use: https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/invoke/MethodHandles.Lookup.html#findClass(java.lang.String)

So instead of a plain `Class.forName()` you can use `LOOKUP.findClass()`, so both fits together much better. This uses the accessibility of the used Lookup and so it ensure that the class is visible and can be loaded. It also simplifies access checks.